mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java

```diff
@@ -1190,42 +1139,103 @@ private void executeBatchRemove(Map<DBCollection, BatchInsertionTask> inserts, R
 		}
 	}
 
-	private void executeBatchUpdate(Map<DBCollection, BatchInsertionTask> inserts, InsertOrUpdateTupleOperation tupleOperation) {
-		EntityKey entityKey = tupleOperation.getEntityKey();
-		Tuple tuple = tupleOperation.getTuple();
-		MongoDBTupleSnapshot snapshot = (MongoDBTupleSnapshot) tupleOperation.getTuple().getSnapshot();
-		WriteConcern writeConcern = getWriteConcern( tupleOperation.getTupleContext() );
-
-		if ( INSERT == snapshot.getSnapshotType() ) {
-			prepareForInsert( inserts, snapshot, entityKey, tuple, writeConcern );
+	private void executeBatchUpdate(Map<DBCollection, BatchInsertionTask> inserts, List<MongoDBTupleSnapshot> insertSnapshots,
```

Comment:
That's the plan. I didn't do it right now because it was easier to follow the flow of operations and easier to debug it.